### PR TITLE
Allow iterating keys on key collection and adopt `swift-log`

### PR DESCRIPTION
**Pull request by ptoffy**
This addresses #169 and fixes a typo.
Also adds `swift-log` as a dependency to avoid `print`ing stuff

### CODE REVIEW

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 1]
```diff
@@ -16,12 +16,20 @@ public actor JWTKeyCollection: Sendable {
     public let defaultJWTParser: any JWTParser
     public let defaultJWTSerializer: any JWTSerializer
 
+    /// Whether the collection should iterate over all signers when verifying a JWT.
+    let shouldIterateKeys: Bool
+
     /// Creates a new empty Signers collection.
     /// - parameters:
     ///    - jsonEncoder: The default JSON encoder.
     ///    - jsonDecoder: The default JSON decoder.
```
Comment (gwynne) [inline Sources/JWTKit/JWTKeyCollection.swift:25]: Add documentation for the new parameter here

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 2]
```diff
@@ -16,12 +16,20 @@ public actor JWTKeyCollection: Sendable {
     public let defaultJWTParser: any JWTParser
     public let defaultJWTSerializer: any JWTSerializer
 
+    /// Whether the collection should iterate over all signers when verifying a JWT.
+    let shouldIterateKeys: Bool
```
Comment (gwynne) [inline Sources/JWTKit/JWTKeyCollection.swift:20]: Which is it, all signers or all keys?

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 3]
```diff
@@ -205,8 +213,26 @@ public actor JWTKeyCollection: Sendable {
...
+                    do {
+                        signer = try self.getSigner(for: kid, alg: header.alg)
+                        return try await signer.verify(token)
+                    } catch {
+                        continue
+                    }
+                }
+
+                throw error
+            }
```
Comment (gwynne) [inline Sources/JWTKit/JWTKeyCollection.swift:234]:
```suggestion
            if self.shouldIterateKeys {
                for (kid, _) in self.storage {
                    do {
                        signer = try self.getSigner(for: kid, alg: header.alg)
                        return try await signer.verify(token)
                    } catch {}
                }
            }
            throw error
```

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 4]
```diff
@@ -38,7 +46,7 @@ public actor JWTKeyCollection: Sendable {
     func add(_ signer: JWTSigner, for kid: JWKIdentifier? = nil) -> Self {
         let signer = JWTSigner(algorithm: signer.algorithm, parser: signer.parser, serializer: signer.serializer)
 
-        if let kid = kid {
+        if let kid {
             if self.storage[kid] != nil {
                 print("Warning: Overwriting existing JWT signer for key identifier '\(kid)'.")
```
Comment (gwynne) [inline Sources/JWTKit/JWTKeyCollection.swift:51]: Is there a `Logger` that can be used here instead of `print()`?

**Review by gwynne — CHANGES_REQUESTED** [review]

File: Sources/JWTKit/JWTKeyCollection.swift (hunk 4 above)
Comment (ptoffy) [reply Sources/JWTKit/JWTKeyCollection.swift:51]: Not really. Adding swift-log for that might not really be worth it

Comment (gwynne) [reply Sources/JWTKit/JWTKeyCollection.swift:52]: Agreed, but at the same time IIRC it's considered bad form for libraries to `print()`. @0xTim Do you know what the recommendation is in such a case?

**Review by gwynne — APPROVED** [review]

File: Sources/JWTKit/JWTKeyCollection.swift (hunk 4 above)
Comment (0xTim) [reply Sources/JWTKit/JWTKeyCollection.swift:52]: Yeah printing is a no go, we need to use a logger if we want to log stuff. I'm surprised we haven't needed to up to this point

**Review by 0xTim — CHANGES_REQUESTED** [review]
Use a logger instead of printing

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 5]
```diff
@@ -205,8 +219,21 @@ public actor JWTKeyCollection: Sendable {
...
-        return try await signer.verify(token)
+        var signer = try self.getSigner(for: kid, alg: header.alg)
+
+        do {
+            return try await signer.verify(token)
+        } catch {
+            if self.shouldIterateKeys {
+                for (kid, _) in self.storage {
+                    do {
+                        signer = try self.getSigner(for: kid, alg: header.alg)
```
Comment (0xTim) [inline Sources/JWTKit/JWTKeyCollection.swift:230]: What happens if a signer doesn't have a `kid`?

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 6]
```diff
@@ -205,8 +219,21 @@ public actor JWTKeyCollection: Sendable {
...
         let header = try defaultJWTParser.parseHeader(token)
         let kid = header.kid.flatMap { JWKIdentifier(string: $0) }
-        let signer = try self.getSigner(for: kid, alg: header.alg)
-        return try await signer.verify(token)
+        var signer = try self.getSigner(for: kid, alg: header.alg)
+
+        do {
+            return try await signer.verify(token)
+        } catch {
+            if self.shouldIterateKeys {
```
Comment (0xTim) [inline Sources/JWTKit/JWTKeyCollection.swift:227]: Minor nit we're verifying the signer twice here

**Review by 0xTim — CHANGES_REQUESTED** [review]
Do we have any other prints?

File: Sources/JWTKit/JWTKeyCollection.swift (hunk 5 above)
Comment (ptoffy) [reply Sources/JWTKit/JWTKeyCollection.swift:230]: It's marked as default. Adding a different signer without a kid will override the one before

**Comment by ptoffy** [comment]
@0xTim no there's no other prints

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 7]
```diff
@@ -38,14 +47,14 @@ public actor JWTKeyCollection: Sendable {
     func add(_ signer: JWTSigner, for kid: JWKIdentifier? = nil) -> Self {
         let signer = JWTSigner(algorithm: signer.algorithm, parser: signer.parser, serializer: signer.serializer)
 
-        if let kid = kid {
+        if let kid {
             if self.storage[kid] != nil {
-                print("Warning: Overwriting existing JWT signer for key identifier '\(kid)'.")
+                logger.warning("Overwriting existing JWT signer", metadata: ["kid": "\(kid)"])
```
Comment (0xTim) [inline Sources/JWTKit/JWTKeyCollection.swift:52]: These should be debug as per the log level guide https://www.swift.org/documentation/server/guides/libraries/log-levels.html

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 8]
```diff
@@ -38,14 +47,14 @@ public actor JWTKeyCollection: Sendable {
...
             }
             self.storage[kid] = .jwt(signer)
         } else {
             if self.default != nil {
-                print("Warning: Overwriting existing default JWT signer.")
+                logger.warning("Overwriting existing default JWT signer")
```
Comment (0xTim) [inline Sources/JWTKit/JWTKeyCollection.swift:57]: As above

**Review by 0xTim — CHANGES_REQUESTED** [review]

File: Sources/JWTKit/JWTKeyCollection.swift [hunk 9]
```diff
@@ -16,14 +17,21 @@ public actor JWTKeyCollection: Sendable {
...
+
     /// Creates a new empty Signers collection.
     /// - parameters:
     ///    - jsonEncoder: The default JSON encoder.
     ///    - jsonDecoder: The default JSON decoder.
-    public init(defaultJWTParser: some JWTParser = DefaultJWTParser(), defaultJWTSerializer: some JWTSerializer = DefaultJWTSerializer()) {
+    public init(
+        defaultJWTParser: some JWTParser = DefaultJWTParser(),
+        defaultJWTSerializer: some JWTSerializer = DefaultJWTSerializer(),
+        logger: Logger = Logger(label: "jwt_kit_do_not_log", factory: { _ in SwiftLogNoOpLogHandler() })
```
Comment (0xTim) [inline Sources/JWTKit/JWTKeyCollection.swift:29]: Sorry I should have added we should have this documented as well

**Review by 0xTim — APPROVED** [review]

**Pull request merged**